### PR TITLE
cooja: remove home-grown thread routines

### LINKED ISSUE #1670: could not run cooja - apple -m1
$> contiker cooja
WARNING: The requested image's platform (linux/386) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Error: A JNI error has occurred, please check your installation and try again
Exception in thread "main" java.lang.ExceptionInInitializerError
	at java.util.Locale.getDefault(Locale.java:836)
	at java.text.MessageFormat.<init>(MessageFormat.java:361)
	at java.text.MessageFormat.format(MessageFormat.java:840)
	at sun.launcher.LauncherHelper.getLocalizedMessage(LauncherHelper.java:426)
	at sun.launcher.LauncherHelper.abort(LauncherHelper.java:547)
	at sun.launcher.LauncherHelper.validateMainClass(LauncherHelper.java:653)
	at sun.launcher.LauncherHelper.checkAndLoadMain(LauncherHelper.java:632)
Caused by: java.lang.ArrayIndexOutOfBoundsException: 0
	at java.util.Locale$1.<clinit>(Locale.java:836)
	... 7 more

### PR DESCRIPTION
**Pull request by pjonsson**
Switch to the POSIX.1-2001 setcontext/getcontext
routines instead of hand-written assembly.
These routines were removed in POSIX.1-2008 for
portability reasons with the recommendation
to use pthreads instead, but that is not feasible
for Cooja.

Fixes #1670

### CODE REVIEW

File: arch/platform/cooja/sys/cooja_mt.c [hunk 1]
```diff
@@ -36,36 +36,95 @@
...
-cooja_mt_start(struct cooja_mt_thread *thread, void (* function)(void *), void *data)
+cooja_mt_start(struct cooja_mt_thread *caller,
+               struct cooja_mt_thread *t, void (*function)(void))
 {
-  /* Call the architecture dependant function to set up the processor
-     stack with the correct parameters. */
-  cooja_mtarch_start(&thread->thread, function, data);
+  for(int i = 0; i < COOJA_MTARCH_STACKSIZE; ++i) {
+    t->stack[i] = i;
+  }
```
Comment (nfi) [inline arch/platform/cooja/sys/cooja_mt.c:102]: What is the reason to initialize the stack with increasing values?

File: arch/platform/cooja/sys/cooja_mt.c [hunk 2]
```diff
@@ -36,36 +36,95 @@
...
+#else
+  pthread_attr_t attrs;
+  if(pthread_getattr_np(self, &attrs) != 0) {
+    return -2;
+  }
+  if(pthread_attr_getstack(&attrs, &stackAddr, &stackSize) != 0) {
+    return -3;
+  }
+#endif
+  getcontext(&t->ctxt);
```
Comment (nfi) [inline arch/platform/cooja/sys/cooja_mt.c:89]: Please check the return value from `getcontext()` for errors, as is done in other places.

File: arch/platform/cooja/sys/cooja_mt.c [hunk 3]
```diff
@@ -36,36 +36,95 @@
...
+
+#include <signal.h>
+#include <stdio.h>
 
 #define MT_STATE_READY   1
 #define MT_STATE_RUNNING 2
 #define MT_STATE_EXITED  5
 
 static struct cooja_mt_thread *current;
+static struct cooja_mt_thread *cooja_running_thread;
```
Comment (nfi) [inline arch/platform/cooja/sys/cooja_mt.c:56]: What is the difference between `current` and `cooja_running_thread`? Are both needed?

File: arch/platform/cooja/sys/cooja_mt.c [hunk 4]
```diff
@@ -36,36 +36,95 @@
...
-     stack with the correct parameters. */
-  cooja_mtarch_start(&thread->thread, function, data);
+  for(int i = 0; i < COOJA_MTARCH_STACKSIZE; ++i) {
+    t->stack[i] = i;
+  }
 
-  thread->state = MT_STATE_READY;
+  if(getcontext(&t->ctxt) == -1) {
+    perror("getcontext failed");
+  }
```
Comment (nfi) [inline arch/platform/cooja/sys/cooja_mt.c:106]: Should we not stop executing if `getcontext()` fails?

File: arch/platform/cooja/mtype.c [hunk 5]
```diff
@@ -55,11 +54,12 @@ intptr_t referenceVar;
 /*
  * Contiki and rtimer threads.
  */
+static struct cooja_mt_thread cooja_thread;
 static struct cooja_mt_thread rtimer_thread;
 static struct cooja_mt_thread process_run_thread;
 /*---------------------------------------------------------------------------*/
 static void
-rtimer_thread_loop(void *data)
+rtimer_thread_loop()
```
[the remark this answers is not in the document]
Comment (pjonsson) [reply arch/platform/cooja/mtype.c]: Fixed.

File: arch/platform/cooja/mtype.c [hunk 6]
```diff
@@ -70,22 +70,25 @@ rtimer_thread_loop(void *data)
 }
 /*---------------------------------------------------------------------------*/
 static void
-process_run_thread_loop(void *data)
+process_run_thread_loop()
```
[the remark this answers is not in the document]
Comment (pjonsson) [reply arch/platform/cooja/mtype.c]: Fixed.

File: arch/platform/cooja/mtype.c [hunk 7]
```diff
@@ -70,22 +70,25 @@ rtimer_thread_loop(void *data)
...
-void
+int
 cooja_init(void)
 {
+  int rv;
   /* Create rtimers and Contiki threads */
-  cooja_mt_start(&rtimer_thread, &rtimer_thread_loop, NULL);
-  cooja_mt_start(&process_run_thread, &process_run_thread_loop, NULL);
+  if((rv = cooja_mt_init(&cooja_thread)))
+    return rv;
```
[the remark this answers is not in the document]
Comment (pjonsson) [reply arch/platform/cooja/mtype.c]: Fixed.

File: arch/platform/cooja/mtype.c [hunk 8]
```diff
@@ -70,22 +70,25 @@ rtimer_thread_loop(void *data)
...
+  cooja_mt_start(&cooja_thread, &rtimer_thread, &rtimer_thread_loop);
+  cooja_mt_start(&cooja_thread, &process_run_thread, &process_run_thread_loop);
```
[the remark this answers is not in the document]
Comment (pjonsson) [reply arch/platform/cooja/mtype.c]: Fixed.

File: arch/platform/cooja/sys/cooja_mt.c [hunk 9]
```diff
@@ -36,36 +36,95 @@
...
+cooja_mt_init(struct cooja_mt_thread *t)
+{
+#ifdef __APPLE__
+  /* Apple's makecontext() fails if size is less than MINSIGSTKSZ. */
+  if(COOJA_MTARCH_STACKSIZE < MINSIGSTKSZ) {
+    return -1;
+  }
+#endif
+  void *stackAddr;
+  size_t stackSize;
```
[the remark this answers is not in the document]
Comment (pjonsson) [reply arch/platform/cooja/sys/cooja_mt.c]: Fixed.

File: arch/platform/cooja/sys/cooja_mt.c (hunk 1 above)
Comment (pjonsson) [reply arch/platform/cooja/sys/cooja_mt.c:107]: This is a leftover from the old code. This PR is already a significant change, so I've added a fixme to remove it in the future.

The new code was writing values that were out of range fro the type, so added a modulo for the initialization value.

File: arch/platform/cooja/sys/cooja_mt.c (hunk 2 above)
Comment (pjonsson) [reply arch/platform/cooja/sys/cooja_mt.c]: Fixed.

File: arch/platform/cooja/sys/cooja_mt.c (hunk 4 above)
Comment (pjonsson) [reply arch/platform/cooja/sys/cooja_mt.c:112]: There isn't any error checking/propagation in the old code, but I've added a return.

File: arch/platform/cooja/sys/cooja_mt.c (hunk 3 above)
Comment (pjonsson) [reply arch/platform/cooja/sys/cooja_mt.c:57]: The old code had both, and I'm not sure why. They are not always the same value, so I left both so the PR wouldn't introduce new behaviors.

Comment (nfi) [reply arch/platform/cooja/sys/cooja_mt.c:57]: I think the reason was that the variables were in different files and local in each file. The variables have now moved to the same file and seem to serve the same purpose, but let us leave it for another PR if you like.

**Review by nfi — APPROVED** [review]

**Comment by simonduq** [comment]
Hi!

This was my mistake, very sorry about it, re-opening this PR now.

What happened is the following: I was switching the base branch for this repository https://github.com/wittra/contiki-ng from develop to wittra. But accidentally, I made the change on the wrong repo (this repo) and instead of switching the base branch I renamed it. And somehow github deleted develop and closed all PRs...

Many apologies for this mishap 🙏; I haven't contributed in a while.. but now at least everybody got some notification from me :p

**Pull request merged**